File: src/main/java/cope/nebula/client/feature/module/movement/LongJump.java
package cope.nebula.client.feature.module.movement;

import cope.nebula.client.events.MotionEvent;
import cope.nebula.client.events.MotionUpdateEvent;
import cope.nebula.client.events.MotionUpdateEvent.Era;
import cope.nebula.client.events.PacketEvent;
import cope.nebula.client.events.PacketEvent.Direction;
import cope.nebula.client.feature.module.Module;
import cope.nebula.client.feature.module.ModuleCategory;
import cope.nebula.client.value.Value;
import cope.nebula.util.internal.math.Vec2d;
import cope.nebula.util.world.entity.player.MotionUtil;
import net.minecraft.network.play.client.CPacketPlayer;
import net.minecraft.network.play.server.SPacketPlayerPosLook;
import net.minecraftforge.fml.common.eventhandler.EventPriority;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;

public class LongJump extends Module {
    public LongJump() {
        super("LongJump", ModuleCategory.MOVEMENT, "Jumps long");
    }

    public static final Value<Mode> mode = new Value<>("Mode", Mode.NCP);
    public static final Value<Boolean> waitForDamage = new Value<>("WaitForDamage", false);
    public static final Value<Double> boost = new Value<>("Boost", 10.0, 0.1, 20.0);
    public static final Value<Integer> timeout = new Value<>("Timeout", 4, 0, 10);
    public static final Value<Boolean> disable = new Value<>("Disable", true);

    private int stage = 1;
    private double moveSpeed = 0.0;
    private double distance = 0.0;

    private int timeoutTicks = 0;
    private int airTicks = 0;

    @Override
    protected void onActivated() {
        stage = 1;
        moveSpeed = 0.0;
        distance = 0.0;

        timeoutTicks = 0;
        airTicks = 0;
    }

    @Override
    public void onTick() {
        if (waitForDamage.getValue()) {
            if (mc.player.hurtTime == 0) {
                timeoutTicks = -1;
            } else {
                timeoutTicks = timeout.getValue() + 1;
            }
        }

        ++timeoutTicks;
    }

    @SubscribeEvent(priority = EventPriority.HIGHEST)
    public void onMotionUpdate(MotionUpdateEvent event) {
        if (event.getEra().equals(Era.PRE)) {
            distance = Math.sqrt(Math.pow(mc.player.posX - mc.player.prevPosX, 2) + Math.pow(mc.player.posZ - mc.player.prevPosZ, 2));
        }
    }

    @SubscribeEvent
    public void onPacket(PacketEvent event) {
        if (event.getDirection().equals(Direction.INCOMING)) {
            if (event.getPacket() instanceof SPacketPlayerPosLook && disable.getValue()) {
                disable();
            }
        } else if (event.getDirection().equals(Direction.OUTGOING)) {
            if (event.getPacket() instanceof CPacketPlayer && timeoutTicks < timeout.getValue()) {
                event.setCanceled(true);
            }
        }
    }

    @SubscribeEvent
    public void onMotion(MotionEvent event) {
        if (timeoutTicks < timeout.getValue()) {
            event.setX(mc.player.motionX = 0.0);
            event.setZ(mc.player.motionZ = 0.0);

            return;
        }

        if (mode.getValue().equals(Mode.NCP)) {
            if (mc.player.onGround) {
                stage = 2;
            }

            if (stage == 1) {
                moveSpeed = boost.getValue() * MotionUtil.getBaseNCPSpeed();
            } else if (stage == 2) {
                if (mc.player.onGround && MotionUtil.isMoving()) {
                    double jumpHeight = MotionUtil.getJumpHeight(false);

                    mc.player.motionY = jumpHeight;
                    event.setY(jumpHeight);

                    moveSpeed *= 2.149;
                }
            } else if (stage == 3) {
                double diff = 0.66 * (distance - MotionUtil.getBaseNCPSpeed());
                moveSpeed = distance - diff;
            } else {
                moveSpeed = distance - distance / 159.0;
            }

            moveSpeed = Math.max(moveSpeed, MotionUtil.getBaseNCPSpeed());

            Vec2d vec = MotionUtil.strafe(moveSpeed);

            event.setX(vec.getX());
            event.setZ(vec.getZ());

            ++stage;
        } else if (mode.getValue().equals(Mode.VANILLA)) {
            if (mc.player.onGround && MotionUtil.isMoving()) {
                double jumpHeight = MotionUtil.getJumpHeight(true);

                moveSpeed = MotionUtil.getBaseNCPSpeed();

                mc.player.motionY = jumpHeight;
                event.setY(jumpHeight);
            }

            if (moveSpeed < 0.0) {
                moveSpeed = boost.getValue();
                airTicks = 1;
            }

            if (!mc.player.onGround) {
                double speed = MotionUtil.getBaseNCPSpeed() * boost.getValue();

                ++airTicks;
                if (airTicks > 0) {
                    moveSpeed = speed - (airTicks / 100.0);
                }
            }

            Vec2d vec = MotionUtil.strafe(moveSpeed);

            event.setX(vec.getX());
            event.setZ(vec.getZ());
        }
    }

    public enum Mode {
        NCP,
        VANILLA
    }
}

File: src/main/kotlin/wtf/nebula/client/feature/module/movement/LongJump.kt
package wtf.nebula.client.feature.module.movement

import me.bush.eventbuskotlin.EventListener
import me.bush.eventbuskotlin.listener
import net.minecraft.network.play.client.CPacketPlayer
import net.minecraft.network.play.server.SPacketPlayerPosLook
import org.lwjgl.input.Keyboard
import wtf.nebula.client.Nebula
import wtf.nebula.client.config.setting.Setting
import wtf.nebula.client.event.packet.PacketReceiveEvent
import wtf.nebula.client.event.player.motion.MotionEvent
import wtf.nebula.client.event.player.motion.update.PreMotionUpdate
import wtf.nebula.client.feature.module.Module
import wtf.nebula.client.feature.module.ModuleCategory
import wtf.nebula.util.ext.isMoving
import wtf.nebula.util.motion.MotionUtil
import kotlin.IndexOutOfBoundsException
import kotlin.doubleArrayOf
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.sin
import kotlin.math.sqrt

class LongJump : Module(ModuleCategory.MOVEMENT, "Makes you go long asf aka best way to pussy out a fight LOL") {
    val mode by setting("Mode", Mode.NORMAL)
    val boost by double("Boost", 6.2, 1.0..10.0)

    private var stage = LongJumpStage.COLLIDE
    private var moveSpeed = 0.0
    private var distance = 0.0

    private var airTicks = 0
    private var groundTicks = 0

    override fun getDisplayInfo(): String? {
        return Setting.formatEnum(mode)
    }

    override fun onActivated() {
        super.onActivated()
        stage = LongJumpStage.START
        moveSpeed = 0.0
        distance = 0.0
    }

    override fun onDeactivated() {
        super.onDeactivated()
        Nebula.serverManager.speed = 1.0f

        airTicks = 0
        groundTicks = 0
    }

    @EventListener
    private val motionListener = listener<MotionEvent> {
        if (mode == Mode.NORMAL) {

            when (stage) {
                LongJumpStage.START -> {
                    stage = LongJumpStage.JUMP
                    moveSpeed = boost * MotionUtil.getBaseNCPSpeed() - 0.01
                }

                LongJumpStage.JUMP -> {
                    stage = LongJumpStage.SPEED

                    mc.player.motionY = 0.42
                    it.y = 0.42

                    moveSpeed *= 2.149
                }

                LongJumpStage.SPEED -> {
                    stage = LongJumpStage.COLLIDE

                    val adjusted = 0.66 * (distance - MotionUtil.getBaseNCPSpeed())
                    moveSpeed = distance - adjusted
                }

                LongJumpStage.COLLIDE -> {
                    val boxes = mc.world.getCollisionBoxes(mc.player, mc.player.entityBoundingBox.offset(0.0, mc.player.motionY, 0.0))
                    if (boxes.isNotEmpty() && mc.player.collidedVertically) {
                        stage = LongJumpStage.START
                    }

                    moveSpeed -= moveSpeed / 159.0
                }
            }

            moveSpeed = max(moveSpeed, MotionUtil.getBaseNCPSpeed())

            val strafe = MotionUtil.strafe(moveSpeed)

            if (mc.player.isMoving()) {
                it.x = strafe[0]
                it.z = strafe[1]
            }

            else {
                it.x = 0.0
                it.z = 0.0
            }
        }

        else if (mode == Mode.COWABUNGA) {
            val dir = MotionUtil.getMovement()[2]

            val xDir = cos((dir + 90.0f) * Math.PI / 180.0).toFloat()
            val zDir = sin((dir + 90.0f) * Math.PI / 180.0).toFloat()

            if (!mc.player.collidedVertically) {
                airTicks++
                groundTicks = 0

                when (mc.player.motionY) {
                    -0.07190068807140403 -> mc.player.motionY *= 0.3499999940395355
                    -0.10306193759436909 -> mc.player.motionY *= 0.550000011920929
                    -0.13395038817442878 -> mc.player.motionY *= 0.6700000166893005
                    -0.16635183030382 -> mc.player.motionY *= 0.6899999976158142
                    -0.19088711097794803 -> mc.player.motionY *= 0.7099999785423279
                    -0.21121925191528862 -> mc.player.motionY *= 0.20000000298023224
                    -0.11979897632390576 -> mc.player.motionY *= 0.9300000071525574
                    -0.18758479151225355 -> mc.player.motionY *= 0.7200000286102295
                    -0.21075983825251726 -> mc.player.motionY *= 0.7599999904632568
                }

                if (mc.player.motionY < -0.2 && mc.player.motionY > -0.24) {
                    mc.player.motionY *= 0.7
                }

                if (mc.player.motionY < -0.25 && mc.player.motionY > -0.32) {
                    mc.player.motionY *= 0.8
                }

                if (mc.player.motionY < -0.35 && mc.player.motionY > -0.8) {
                    mc.player.motionY *= 0.98
                }

                if (mc.player.motionY < -0.8 && mc.player.motionY > -1.6) {
                    mc.player.motionY *= 0.99
                }

                Nebula.serverManager.speed = 0.8f

                if (Keyboard.isKeyDown(mc.gameSettings.keyBindForward.keyCode)) {
                    try {
                        val speedVal = AIR_SPEED[airTicks - 1] * 3.0
                        mc.player.motionX = xDir * speedVal
                        mc.player.motionZ = zDir * speedVal
                    } catch (ignored: IndexOutOfBoundsException) {
                    }
                }

                else {
                    mc.player.motionX = 0.0
                    mc.player.motionZ = 0.0
                }

            }

            else {
                Nebula.serverManager.speed = 1.0f

                airTicks = 0
                groundTicks++

                mc.player.motionX /= 13.0
                mc.player.motionZ /= 13.0

                if (groundTicks == 1) {
                    sendPacketPos(0.0, 0.0, 0.0)
                    sendPacketPos(0.0624, 0.0, 0.0)
                    sendPacketPos(0.0, 0.419, 0.0)
                    sendPacketPos(0.0624, 0.0, 0.0)
                    sendPacketPos(0.0, 0.419, 0.0)
                }

                if (groundTicks > 2) {
                    groundTicks = 0

                    mc.player.motionX = xDir * 0.3
                    mc.player.motionZ = zDir * 0.3
                    mc.player.motionY = 0.42399999499320984
                }
            }

            it.x = mc.player.motionX
            it.y = mc.player.motionY
            it.z = mc.player.motionZ
        }
    }

    @EventListener
    private val preMotionUpdate = listener<PreMotionUpdate> {
        val diffX = mc.player.posX - mc.player.prevPosX
        val diffZ = mc.player.posZ - mc.player.prevPosZ

        distance = sqrt(diffX * diffX + diffZ * diffZ)
    }

    @EventListener
    private val packetReceiveListener = listener<PacketReceiveEvent> {
        if (it.packet is SPacketPlayerPosLook) {
            toggled = false
        }
    }

    private fun sendPacketPos(x: Double, y: Double, z: Double) {
        mc.player.connection.sendPacket(CPacketPlayer.Position(x, y, z, false))
    }

    enum class Mode {
        NORMAL, COWABUNGA
    }

    enum class LongJumpStage {
        START, JUMP, SPEED, COLLIDE
    }

    companion object {
        private val AIR_SPEED = doubleArrayOf(
            0.420606, 0.417924, 0.415258, 0.412609, 0.409977, 0.407361, 0.404761, 0.402178, 0.399611, 0.39706,
            0.394525, 0.392, 0.3894, 0.38644, 0.383655, 0.381105, 0.37867, 0.37625, 0.37384, 0.37145,
            0.369, 0.3666, 0.3642, 0.3618, 0.35945, 0.357, 0.354, 0.351, 0.348, 0.345,
            0.342, 0.339, 0.336, 0.333, 0.33, 0.327, 0.324, 0.321, 0.318, 0.315,
            0.312, 0.309, 0.307, 0.305, 0.303, 0.3, 0.297, 0.295, 0.293, 0.291,
            0.289, 0.287, 0.285, 0.283, 0.281, 0.279, 0.277, 0.275, 0.273, 0.271,
            0.269, 0.267, 0.265, 0.263, 0.261, 0.259, 0.257, 0.255, 0.253, 0.251,
            0.249, 0.247, 0.245, 0.243, 0.241, 0.239, 0.237
        )
    }
}